File: js/src/test/scala/NameSpace.scala
import org.scalatest.funsuite.AnyFunSuite
import ts2mls.TSProgram
import ts2mls.types._

class Namespace extends AnyFunSuite {
  test("Namespace") {
    val program = TSProgram(Namespace.testFiles)
    val ns = program.>>("N1")
    assert(TSTypeTest(ns.>("f"), "(any) => number"))
    assert(TSTypeTest(ns.>("C").>("f"), "void"))
    assert(TSTypeTest(ns.>>("N2").>("fff"), "(boolean) => number"))
    assert(TSTypeTest(ns.>>("N2").>("f"), "(any) => number"))
  }
}

object Namespace {
  private val testFiles = TSTypeTest.tsPathes(Seq("NameSpace.ts"))
}

File: js/src/test/typescript/NameSpace.ts
namespace N1 {
  export function f(x) {
    return 42;
  }

  function ff(y) {
    return 42;
  }

  export class C {
    f() {}
  }

  interface I {
    f: () => number
  }

  export namespace N2 {
    /**
     * @debug
     */
    export function fff(x: boolean) {
      return 42;
    }
  }
}

namespace AA {
  export function f(x) {
    return "42";
  }

  export class C {
    f() {}
  }

  export interface I {
    f: () => number
  }

  export namespace N2 {}
}

/**
  * @debug
  */
function f1(x: N1.C): N1.C {
  return x;
}

/**
  * @debug
  */
function f2(x: AA.C): AA.C {
  return x;
}